store/db.go

```diff
@@ -29,9 +30,42 @@ const (
 	DbFilename = "ql.db"
 )
 
+// dbLock is used to guarantee both thread-safety and process-safety
+// for db access.
+type dbLock struct {
+	fl *lock.FileLock
+	// Although the FileLock already ensures thread safety, the Go runtime is unaware
+	// of this, and so Mutex is necessary to satisfy the Go race detector
+	sync.Mutex
+}
+
+func newDBLock(dirPath string) (*dbLock, error) {
+	l, err := lock.NewLock(dirPath, lock.Dir)
+	if err != nil {
+		return nil, err
+	}
+	return &dbLock{fl: l}, nil
+}
+
+func (dl *dbLock) lock() error {
+	if err := dl.fl.ExclusiveLock(); err != nil {
+		return err
+	}
+	dl.Lock()
+	return nil
+}
+
+func (dl *dbLock) unlock() error {
+	if err := dl.fl.Unlock(); err != nil {
+		return err
+	}
+	dl.Unlock()
```

Comment:
How about we move the unlocking of the mutex above unlocking of the file lock? That way we have a LIFO unlocking, and we always unlock the locked mutex, even if file unlock fails.